File: sofia-core/src/main/java/com/antwerkz/sofia/SofiaConfig.java
package com.antwerkz.sofia;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Properties;
import java.util.TreeMap;

public class SofiaConfig {
  private String bundleName = "sofia";
  private String className;
  private String packageName = "com.antwerkz.sofia";
  private File outputDirectory;
  private LoggingType type = LoggingType.NONE;
  private boolean useControl = false;
  private Map<String, String> properties;
  private boolean generateJavascript;
  private File javascriptOutputFile;

  public String getBundleName() {
    return bundleName;
  }

  public String getClassName() {
    return className;
  }

  public File getOutputDirectory() {
    return outputDirectory;
  }

  public String getPackageName() {
    return packageName;
  }

  public Map<String, String> getProperties() {
    return properties;
  }

  public LoggingType getType() {
    return type;
  }

  public boolean isUseControl() {
    return useControl;
  }

  public SofiaConfig setBundleName(String bundleName) {
    this.bundleName = bundleName;
    return this;
  }

  public SofiaConfig setClassName(String className) {
    this.className = className;
    return this;
  }

  public SofiaConfig setOutputDirectory(File outputDirectory) {
    this.outputDirectory = outputDirectory;
    return this;
  }

  public SofiaConfig setOutputDirectory(String outputDirectory) {
    this.outputDirectory = new File(outputDirectory);
    return this;
  }

  public SofiaConfig setPackageName(String packageName) {
    this.packageName = packageName;
    return this;
  }

  public SofiaConfig setProperties(File propertiesFile) {
    try {
      this.properties = loadProperties(new FileInputStream(propertiesFile));
    } catch (FileNotFoundException e) {
      throw new RuntimeException(e.getMessage(), e);
    }
    return this;
  }

  public SofiaConfig setProperties(Map<String, String> properties) {
    this.properties = properties;
    return this;
  }

  public SofiaConfig setProperties(InputStream stream) {
    this.properties = loadProperties(stream);
    return this;
  }

  public SofiaConfig setUseControl(boolean useControl) {
    this.useControl = useControl;
    return this;
  }

  public SofiaConfig setType(LoggingType type) {
    this.type = type;
    return this;
  }

  private Map<String, String> loadProperties(InputStream inputStream) {
    try (InputStream stream = inputStream) {
      properties = new TreeMap<>();
      Properties props = new Properties();
      props.load(stream);
      for (Entry<Object, Object> entry : props.entrySet()) {
        properties.put((String) entry.getKey(), (String) entry.getValue());
      }
    } catch (IOException e) {
      throw new RuntimeException(e.getMessage(), e);
    }
    return properties;
  }

  public boolean isGenerateJavascript() {
    return generateJavascript;
  }

  public SofiaConfig setGenerateJavascript(boolean generateJavascript) {
    this.generateJavascript = generateJavascript;
    return this;
  }

  public File getJavascriptOutputFile() {
    return javascriptOutputFile;
  }

  public SofiaConfig setJavascriptOutputFile(File file) {
    this.javascriptOutputFile = file;
    return this;
  }
}

File: sofia-core/src/main/kotlin/com/antwerkz/sofia/SofiaConfig.kt
package com.antwerkz.sofia

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.Charset
import java.util.ArrayList
import java.util.Locale
import java.util.Properties
import java.util.TreeMap

class SofiaConfig(
        val propertiesFile: File,
        var outputDirectory: File,
        var loggingType: LoggingType = LoggingType.NONE,
        var packageName: String = "com.antwerkz.sofia",
        var generateJava: Boolean = true,
        var generateKotlin: Boolean = false,
        var bundleName: String = propertiesFile.nameWithoutExtension,
        var className: String = bundleName.capitalize(),
        var charset: Charset = Charset.forName("ISO-8859-1")) {

    var bundles: MutableMap<String, List<Method>> = sortedMapOf()
    val methods: List<Method>
    private val language: String = when {
        generateJava -> "java"
        generateKotlin -> "kotlin"
        else -> throw RuntimeException("Either 'java' or 'kotlin' must be selected to generate")
    }

    init {
        methods = mapMethods(loadProperties(propertiesFile))
        discoverBundles(propertiesFile)
    }

    fun findParentExtension(bundle: String): String {
        var parent = bundle
        do {
            parent = parent.split("_").dropLast(1).joinToString("_")
        } while (bundles[parent] == null && parent != "")
        return if (parent == "") parent else "_$parent"
    }

    private fun mapMethods(map: MutableMap<String, String>): ArrayList<Method> {
        val list = ArrayList<Method>()
        for (entry in map.entries) {
            val method = Method(language, loggingType, entry.key, entry.value)
            list.add(method)
            if (method.logged && !loggingType.loggingLevels.contains(method.logLevel!!)) {
                throw IllegalArgumentException("Invalid logging level '${method.logLevel}' for logging type '$loggingType'")
            }
        }
        return list
    }

    private fun discoverBundles(propertiesFile: File) {
        val name = propertiesFile.name
        val rootDir = propertiesFile.parent
        val last = name.lastIndexOf('.')
        if (!name.matches(".*_[a-zA-Z]+[_a-zA-Z]*.properties".toRegex())) {
            for (locale in Locale.getAvailableLocales()) {
                val file = File(rootDir, "%s_%s%s".format(name.substring(0, last), locale.toString(), name.substring(last)))
                if (file.exists()) {
                    bundles.put(locale.toString(), mapMethods(loadProperties(file)))
                }
            }
        }
    }

    private fun loadProperties(file: File): MutableMap<String, String> {
        val map = TreeMap<String, String>()
        try {
            InputStreamReader(FileInputStream(file), charset).use { reader ->
                val props = Properties()
                props.load(reader)
                for (entry in props.entries) {
                    map.put(entry.key as String, entry.value as String)
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e.message, e)
        }

        if (map.isEmpty()) {
            throw RuntimeException("No properties found in properties file.")
        }
        return map
    }
}
